### PR TITLE
Example using Quantum MDM on real data

### PR DESCRIPTION
**Pull request by toncho11**
This PR is focused on testing the Quantum MDM on real data. 
The process will take some time to validate the code and the results.

### CODE REVIEW

**Comment by gcattan** [comment]
Hi, very good. In addition,  think we can test for different couples of distance and mean for both MDM and QuanticMDM.
I agree this will take some times. Let's try with `Quantum=False` in the first time as you suggested, and then do the validation with `Quantum=True` in a second PR.

Do you want me to take a look at the flake8 error?

**Comment by toncho11** [comment]
Can we disable it temporally for now? :)

Yes, there are a number of parameters to test:

- Covariance, ErpCovariance, XDawnCovariance
- number of filters for the XDawn Covariance 
- ...

**Comment by toncho11** [comment]
List of things to test in order to evaluate the new algorithm. 

1) WithinSubjectEvaluation

2) `Shrinkage(shrinkage=0.01)` and `Shrinkage(shrinkage=0.8)` ?

3) In docplex.py,we can try also to change `CobylaOptimizer(rhobeg=0.01, rhoend=0.0001)` to `CobylaOptimizer(rhobeg=0.1, rhoend=0.00001)`

4) mean = logeuclid, distance = convex , check y_pred if it is all 0 or all 1s. 

5) test merging all subjects and splitting in train/validate to see if the algorithm we generalize better on many subjects instead of per session.

**Comment by gcattan** [comment]
3. it has an importnace. Here with rho in the range 1...0.000001
image

**Comment by gcattan** [comment]
There is still some issue with the regularization of the convex mean, so you can only use mean=convex with distance=euclid.
It does help to change rho and shrinkage as follows:
rhobeg=2.1, rhoend=0.000001, shrinkage=0.9

**Comment by gcattan** [comment]
The mean returned by the docplex model still returns some zeros. It is a problem for other distances than the euclidian one.

I tried the following:
- ignore NaN when getting the log of the matrix
- use a warm start 
- replace zeros by some default values

It run, but the results were worst. @qbarthelemy do you have an idea?

File: examples/ERP/classify_P300_bi_quantum_mdm.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+# if executed on the entire dataset
+n_subjects = 5
+for dataset in datasets:
+    dataset.subject_list = dataset.subject_list[0:n_subjects]
+
+overwrite = True  # set to True if we want to overwrite cached results
+pipelines = {}
+
+pipelines["Xdawn+QuanticMDM"] = make_pipeline(
+    XdawnCovariances(nfilter=8, estimator="lwf", xdawn_estimator="scm"),
```
[the remark this answers is not in the document]
Comment (gcattan) [reply examples/ERP/classify_P300_bi_quantum_mdm.py:108]: Yes, thanks! It gives good results :)

File: pyriemann_qiskit/utils/distance.py [hunk 2]
```diff
@@ -41,7 +42,7 @@ def logeucl_dist_convex(X, y, optimizer=ClassicalOptimizer()):
     classes = range(n_classes)
 
     def dist(m1, m2):
-        return distance_logeuclid(m1, m2)
+        return np.nansum(logm(m1).flatten() * logm(m2).flatten())
```
[the remark this answers is not in the document]
Comment (gcattan) [reply pyriemann_qiskit/utils/distance.py:45]: "dist" might be a little confusing. What I tried to do, is to compute the term `log(A) . log(B)` as per the screenshot below.

image

**Comment by gcattan** [comment]
Here are some of the results.

mean=convex/dist=convex run without error with Xdawn covariances. The results are not good, but also it is not the same space.

In blue is a VotingClassifier with mean=convex/dist=euclid and mean=logeuclid/dist=convex.


Figure_1

**Comment by toncho11** [comment]
> Here are some of the results.
> 
> mean=convex/dist=convex run without error with Xdawn covariances (with corrected estimators as suggested above). The results are not good, but also it is not the same space.
> 
> In blue is a VotingClassifier with mean=convex/dist=euclid and mean=logeuclid/dist=convex.
> 
> Figure_1

So the above results are with BNCI2014009().

**Comment by gcattan** [comment]
Result with one subject, with quantum=True.

mean=convex/distance=euclid

image

mean=logeuclid/distance=convex

image


With mean=convex, the size of the matrix to optimize is too big:

image


So you have to modify the pipeline as follows:
- reduce the xdawn filter to 1 components. This will reduce the size of the covariance matrices to 4x4
- Add an additional Whitening with a dimension reduction (2 components in the figure above)
- In quantic MDM, reduce the upper-bound of the NaiveQAOAOptimizer (4 in the figure above).

**Comment by gcattan** [comment]
Results with one subject and quantum simulation.

Voting classifier is doing a little bit better.

Figure_1

image

**Review by gcattan — APPROVED** [review]

**Pull request merged**